Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -294,17 +305,36 @@ struct PathHierarchy {
     }
     
     private func findNode(path rawPath: String, parent: ResolvedIdentifier?, onlyFindSymbols: Bool) throws -> Node {
-        // The search for a documentation element can be though of as 3 steps:
         // First, parse the path into structured path components.
         let (path, isAbsolute) = Self.parse(path: rawPath)
         guard !path.isEmpty else {
             throw Error.notFound(availableChildren: [])
         }
         
-        // Second, find the node to start the search relative to.
-        // This may consume or or more path components. See implementation for details.
-        var remaining = path[...]
-        var node = try findRoot(parentID: parent, remaining: &remaining, isAbsolute: isAbsolute, onlyFindSymbols: onlyFindSymbols)
+        // Second, we try to the node matching the path. This is done by first finding
+        // the root of the (possibly relative) path and then searching for the child
+        // from that root.
+        // A relative path could have multiple root candidates (where the first
+        // component is a match). We start searching at the `parent`, working
+        // our way up the tree, trying to find a child for each root candidate.
+        // This function reports all errors found on the way and - if successful - the
+        // matching node.
+        let (node, errors) = try searchForChildOnAllPossibleRoots(parentID: parent, path: path, isAbsolute: isAbsolute, onlyFindSymbols: onlyFindSymbols)
```

Comment:
> If I understand this change it's updating the implementation to be branching so that if the node can't be found in the first root match it looks in the second root match and so on. It that correct?

Yes.

> If so, can you describe why this behavior is needed? It would be both be useful to know for the review but also it's be useful information to add in the test.

This is required in the context of extensions to external types in the following scenario:

Here's the [(relevant test)](https://github.com/apple/swift-docc/blob/afe3c497841cdc573a832023fcddd4a222ae5bd8/Tests/SwiftDocCTests/Infrastructure/ReferenceResolverTests.swift#L374) and the relevant excerpt from the bundle used in that test:
```md
#### `/Dependency/UnambiguousType`

It should be possible to reference `/Dependency/UnambiguousType` relatively, even from `documentation/BundleWithRelativePathAmbiguity`. This way, the relative link switches to the extended type page automatically when the type gets extended. 

- ``Dependency/UnambiguousType`` (relative)
- ``/Dependency/UnambiguousType`` (absolute, shorthand)
- ``doc://org.swift.docc.example/documentation/Dependency/UnambiguousType`` (absolute, fully qualified)
```

The problem is that - without the change - the algorithm would try to resolve ` ``Dependency/UnambiguousType`` ` (i.e. the relative syntax) to `documentation/BundleWithRelativePathAmbiguity/Dependency/UnambiguousType`, which does not exist.

The original algorithm matches `Dependency` to `documentation/BundleWithRelativePathAmbiguity/Dependency` and locks in on that decision before it discovers that `documentation/BundleWithRelativePathAmbiguity/Dependency` has no child called `UnambiguousType` and thus fails.

Now, the algorithm just backtracks to the point where it tried the match with `documentation/BundleWithRelativePathAmbiguity/Dependency` and explores root-matches further up the tree, where it finds the correct `Dependency` as a child of `documentation/`.

However, this change also has consequences in other contexts:

Consider the following example:
```swift
struct Identifier {
    func foo() { }
}

struct Kind {
    struct Identifier {
        /// ``Identifier/foo()`` can be resolved by the new implementation whereas the
        /// original algorithm would fail as `Kind.Idenitfier` has no child called `foo()`.
        func bar() { }
    }
}
```

(It is up to discussion whether or not this behavior is considered beneficial or not.) If not, I could also restrict the backtracking to only work for external modules, i.e. the case relevant to extensions to external types.